### ISSUE TITLE
pascals-triangle: adjust API to use an instance method over class methods

### ISSUE THREAD

**Issue by jtigger** [exercism/xjava#365]
As per #177, adjust the test suite to require a solution that uses an instance method instead of a class method.

_(see #344 as an example and the conversation about Function Object vs. Value Object)_

**Comment by stkent** [comment]
Currently this exercises requires a single class that contains two methods with the following signatures:

    public static int[][] computeTriangle(int)

    public static boolean isTriangle(int[][])

Thoughts as to whether these should remain in a single class after conversion to instance methods? The responsibilities of each method are quite different, so I was thinking that two separate classes might be more appropriate here...

**Comment by jtigger** [comment]
I absolutely agree, Stuart.

`computeTriangle()` is a generator a representation of a Pascal's Triangle whereas `isTriangle()` is a validator.

Also, seems like the latter is dependent on the former (or I guess could be?).  This is the first opportunity I've faced with considering dependency injection on the track.  Don't worry, I'm not talking about depending on Dagger or Spring or anything... :)

The question in my mind is whether the `TriangleValidator` should depend on a `TriangleGenerator`?

...

Also... until now, I'm pretty sure we haven't required the practitioner to submit more than one file.  Should we should include a `HINT.md` to "remind" them of that feature of the CLI exists?

**Comment by stkent** [comment]
> Dagger or Spring

😰 😂 

> The question in my mind is whether the `TriangleValidator` should depend on a `TriangleGenerator`?

Hmm... I think it's reasonable here. I'm assuming you're thinking of constructor injection - that would seem most natural to me based on how we'd specify the tests. And yes, if we go this route, a `HINT.md` file is definitely in order. In that case I'd also like to use this as an opportunity to begin documenting these sorts of policies in a single centralized location so that it's easier to regularly review the track for compliance with said policies. In this case, the policy statement would be something like:

> The first exercise in the track whose test suite mandates multiple solution files should be accompanied by a `HINT.md` file reminding the practitioner that the CLI supports multiple file submissions.

and the description of situations in which we would want to double-check the application of this policy would be:

> Any time the track order changes/any time the number of solution files required by an exercise test suite changes.

[If these events weren't so infrequent, I'd suggest that we could build this checklist into a [pull request template](https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/), but I think it would add too much noise.]

**Comment by stkent** [comment]
tl;dr: I'll have a crack at all of the above and we can then review and see how it feels!

**Comment by jtigger** [comment]
@stkent said:
> In that case I'd also like to use this as an opportunity to begin documenting these sorts of policies in a single centralized location so that it's easier to regularly review the track for compliance with said policies. In this case, the policy statement would be something like:
> > The first exercise in the track whose test suite mandates multiple solution files should be accompanied by a HINT.md file reminding the practitioner that the CLI supports multiple file submissions.

I couldn't agree more on both the notion of (finally) standing-up a policy page (perhaps organized by triggering event) and the way you've articulated _this_ particular one.

**Comment by stkent** [comment]
Turns out the canonical data removes the validation steps from this exercise! I updated the implementation to match while addressing this issue. I'll make a new issue to kick off policy tracking!
